### ISSUE TITLE
Refactoring the frontend code

### ISSUE THREAD

**Issue by typerqeo** [Miodec/monkeytype#495]
## Problems
1. Some files are large. `script.js` is especially massive at >5000 LOC.
1. The code is written in a mostly procedural style and has many globals. This makes it hard to follow and error-prone. There is also a lack of tests.
1. External dependencies are currently managed manually.

## Proposed Solutions
1. Let's keep files reasonably small and self-contained. We can use [Webpack](https://webpack.js.org/) to bundle the resources. This would be much nicer for both code organization and deployment.
1. Let's write new JS files in ES6 object-oriented style.
    - This makes it easier to write unit tests and more modular components.
    - The transition doesn't have to take place all at once. The code can be refactored one feature at a time. We should make sure to add unit tests at each step.
    - This might break some functionality in the short term but is undoubtedly good for code health in the long term. We may need to work on another branch for a while.
1. Now that there is a `package.json`, we can use npm to manage dependencies.

---

I'll try to put together a PR to kick off this effort, but I might not have the time to personally see it through to completion. Judging from the hours I've put into this already, this plan is pretty ambitious.

Help or feedback is welcome and appreciated!

**Comment by Miodec** [comment]
Simmilar to #493, not sure which one to close

**Comment by typerqeo** [comment]
Oh... maybe we should repurpose #493 to address the backend? The cloud functions can be dealt with independently from the frontend.

**Comment by Miodec** [comment]
Yeah, sounds good

**Comment by typerqeo** [comment]
Feel free to assign this to me.

**Comment by ivanjermakov** [comment]
+1
I would like to contribute, if the code was cleaner.

**Comment by Marco-Pagani** [comment]
seconded, I would like to help with the refactoring too. @typerqeo once you create a bit of code to establish some conventions feel free to delegate some work to me and I'll try to keep it consistent.

**Comment by typerqeo** [comment]
Work done so far (because I forgot to tag them here earlier):
https://github.com/Miodec/monkeytype/pull/483 https://github.com/Miodec/monkeytype/pull/522 https://github.com/Miodec/monkeytype/pull/526 https://github.com/Miodec/monkeytype/pull/553

**Comment by typerqeo** [comment]
With the above two PRs merged, the stage will be set for

1. converting internal dependencies to ES6 modules
2. converting external dependencies to npm

If you'd like to take on any work converting dependencies, feel free to announce it here. After these PRs are merged, I'll do some more converting if no one's signed up.

The current goal is to get to a point where we don't have to concatenate JS files anymore.

**Comment by aditya1rawat** [comment]
There's a package that alllows you to check for unused dependencies: [Depcheck](https://www.npmjs.com/package/depcheck)

**Comment by thescripted** [comment]
I can work through updating some of the older files today. Additionally, It's may be useful to note PRs to #583 should only fix already refactored files to avoid working on files that will need to be updated anyway.

**Comment by aditya1rawat** [comment]
i haven't used dependabot a whole lot, but does it actually help, or is it not really useful?

**Comment by LeSirH** [comment]
> +1
> I would contribute, if the code was cleaner.

@ivanjermakov Agreed. 

I'm curious how progress has been going with this and how I can potentially help.

**Comment by Miodec** [comment]
> > +1
> 
> > I would contribute, if the code was cleaner.
> 
> 
> 
> @ivanjermakov Agreed. 
> 
> 
> 
> I'm curious how progress has been going with this and how I can potentially help.
> 
> 

Yeah i mean I'm not really sure either since the files are so big it seems like only i would be the only one who can unwind this mess. I even tried to visualise out somehow by going through the files one by one and maaay have some ideas.

https://user-images.githubusercontent.com/13181393/110581220-04aa9b80-8162-11eb-8ec9-ccfe4d26afb9.png

**Comment by Miodec** [comment]
Alright, small step complete. I think once I get all the typing test code related code into a module it should be fine after that (I think most of the issues come from calling `restartTest()` in other files)

**Comment by Miodec** [comment]
![itsdone](https://media0.giphy.com/media/zCME2Cd20Czvy/giphy.gif?cid=ecf05e477r5i3hxfex6lbx0hfoxiyk7wx9jgfoieo8zxgxwt&rid=giphy.gif)



That took a while. I think it still could use a better structure (folder wise), and some files could be broken up more, but everything is a module and allows us to move forward. Hopefully, @typerqeo is still around, and can lead this revolution once more, because I still have no clue how webpacks work and stuff. Thanks for forcing me to rewrite it all into a more readable format.
